### ISSUE TITLE
3com comware v3/H3C comware v5 oxidized not working

### ISSUE THREAD

**Issue by sajiby3k** [ytti/oxidized#1723]
Hi,

Oxidized does not work for 3com comware v3 and H3C comware v5.
It is exactly the same error.

These 3com does not have any command named 'screen-length disable'.

The only way to get that functionality is -
system-view
  user-interface vty 0 4
     screen-length 0

I know I can send those commands via post-login and restore them again in pre logout.

But the problem is it never finds the prompt and I am getting `send': closed stream (IOError).

I have disabled - post_login 'screen-length disable' in model file.

Then tried to run a simple command via oxidized script to see what happens.

From log in the switch I can see that, ssh login successful, then 'undo terminal monitor' is executed. But after that no other command is issued to the switch then the ssh user logs out.

I am using the latest version of oxidized. 

Below is the output can somebody have a look and tell me what to do - 

-bash-4.2$ oxs -d 10.10.10.10 'display lldp neighbor-information brief'
D, [2019-03-05T10:17:36.441581 #126131] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to 10.10.10.10
D, [2019-03-05T10:17:36.441643 #126131] DEBUG -- : AUTH METHODS::["none", "publickey", "password", "keyboard-interactive"]
D, [2019-03-05T10:17:40.271261 #126131] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^\0*(<[\w.-]+>).?$/] at 10.10.10.10
D, [2019-03-05T10:17:41.278030 #126131] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at 10.10.10.10
**D, [2019-03-05T10:17:41.278127 #126131] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: "undo terminal monitor", block: nil at 10.10.10.10**
**D, [2019-03-05T10:17:41.278157 #126131] DEBUG -- : lib/oxidized/input/ssh.rb undo terminal monitor @ 192.168.48.10 with expect: /^\0*(<[\w.-]+>).?$/**
D, [2019-03-05T10:17:41.278333 #126131] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^\0*(<[\w.-]+>).?$/] at 10.10.10.10
**D, [2019-03-05T10:17:42.987657 #126131] DEBUG -- : lib/oxidized/model/model.rb Executing display lldp neighbor-information brief**
D, [2019-03-05T10:17:42.987774 #126131] DEBUG -- : lib/oxidized/input/ssh.rb display lldp neighbor-information brief @ 10.10.10.10 with expect: /^\0*(<[\w.-]+>).?$/
**D, [2019-03-05T10:17:42.988069 #126131] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^\0*(<[\w.-]+>).?$/] at 10.10.10.10**
**closed stream**
Traceback (most recent call last):
        27: from /usr/local/rvm/gems/ruby-2.6.0/bin/ruby_executable_hooks:24:in `<main>'
        26: from /usr/local/rvm/gems/ruby-2.6.0/bin/ruby_executable_hooks:24:in `eval'
        25: from /usr/local/rvm/gems/ruby-2.6.0/bin/oxs:23:in `<main>'
        24: from /usr/local/rvm/gems/ruby-2.6.0/bin/oxs:23:in `load'
        23: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-script-0.6.0/bin/oxs:5:in `<top (required)>'
        22: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-script-0.6.0/lib/oxidized/script/cli.rb:41:in `run'
        21: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-script-0.6.0/lib/oxidized/script/script.rb:23:in `cmd'
        20: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-0.25.1/lib/oxidized/model/model.rb:104:in `cmd'
        19: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-0.25.1/lib/oxidized/input/ssh.rb:48:in `cmd'
        18: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-0.25.1/lib/oxidized/input/ssh.rb:104:in `cmd_shell'
        17: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-0.25.1/lib/oxidized/input/ssh.rb:111:in `expect'
        16: from /usr/local/rvm/rubies/ruby-2.6.0/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
        15: from /usr/local/rvm/rubies/ruby-2.6.0/lib/ruby/2.6.0/timeout.rb:33:in `catch'
        14: from /usr/local/rvm/rubies/ruby-2.6.0/lib/ruby/2.6.0/timeout.rb:33:in `catch'
        13: from /usr/local/rvm/rubies/ruby-2.6.0/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
        12: from /usr/local/rvm/rubies/ruby-2.6.0/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
        11: from /usr/local/rvm/gems/ruby-2.6.0/gems/oxidized-0.25.1/lib/oxidized/input/ssh.rb:112:in `block in expect'
        10: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/session.rb:178:in `loop'
         9: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/session.rb:178:in `loop'
         8: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/session.rb:178:in `block in loop'
         7: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/session.rb:216:in `process'
         6: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/event_loop.rb:29:in `process'
         5: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/event_loop.rb:110:in `ev_select_and_postprocess'
         4: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/session.rb:270:in `ev_do_handle_events'
         3: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/session.rb:270:in `each'
         2: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/connection/session.rb:271:in `block in ev_do_handle_events'
         1: from /usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/buffered_io.rb:102:in `send_pending'
**/usr/local/rvm/gems/ruby-2.6.0/gems/net-ssh-4.1.0/lib/net/ssh/buffered_io.rb:102:in **`send': closed stream (IOError)****

**Comment by berilslv** [comment]
@ config have you tried to add:
comware_cmdline:
enable: 
"comware_cmdline" helpe me with HPE1920

**Comment by sajiby3k** [comment]
Hi,

> cfg :telnet, :ssh do
>     # enable command-line mode on SMB comware switches (HP V1910, V1920)
>     # autodetection is hard, because the 'summary' command is paged, and
>     # the pager cannot be disabled before _cmdline-mode on.
>     if vars :comware_cmdline
>       post_login do
>         send "_cmdline-mode on\n"
>         send "y\n"
>         send vars(:comware_cmdline) + "\n"
>         send "xtd-cli-mode on\n"
>         send "y\n"
>         send vars(:comware_cmdline) + "\n"
>       end
>     end

It does not apply to my switches. Mine are 3COM/H3C, Not HP 1910/1920.

**Comment by sajiby3k** [comment]
At last found solution.

Let's talk about H3C first.

In my oxidized config file, I have globally enabled -

> vars:
>    ssh_no_keepalive: true

Without this H3C, in my environment does not work properly.

Now 3com, they also needs the above "ssh_no_keepalive: true", as this step is enabled globally I do not need to do that for 3com seperately.

But as the 3com, do not support  'screen-length disable', I have extended the default comware.rb model with a new model named "Thcom" to tackle this situation -

In the new model file I have enabled some lines which are commented in the original comware.rb file -

>   expect /^\s*---- More ----$/ do |data, re|
>     send ' '
>     data.sub re, ''
>   end
> 
>   cmd :all do |cfg|
>     # example how to handle pager
>     cfg.gsub! /^.*\e\[42D/, ''
>     # skip rogue ^M
>     cfg = cfg.delete "\r"
>     cfg.cut_both
>   end 

Those lines gives me the opportunity to capture the configuration, even when pager is enabled. Because those lines detects the pager.

More info is available at - 

https://github.com/ytti/oxidized/blob/master/docs/Creating-Models.md
https://github.com/ytti/oxidized/blob/master/docs/Configuration.md

Nice work by oxidized team and I hope my information helps someone else.
